### PR TITLE
#72 - Split approximation option for decomposition and analysis

### LINKED ISSUE #72: Split approximation option for decomposition and analysis
Currently we use `iterative_refinement` for both the approximation of `X0` and of the other sets.
But the precision of `X0` has a major impact of the later precision, so we may want to use different settings there, even different values of `ε`.

### PR DESCRIPTION
**Pull request by schillic**
Closes #72.

* [x] remove redundant `iterative_refinement` option
* [x] use Hyperrectangle as default set type if `ɛ == Inf`
* [x] add separate options for decomposition and approximation

### CODE REVIEW

**Comment by mforets** [comment]
looks good 👍

**Comment by schillic** [comment]
Damn, there are random tests with random errors. I will have a look.

File: src/ReachSets/project_reach.jl [hunk 1]
```diff
@@ -76,21 +74,26 @@ function project_reach(plot_vars::Vector{Int64}, n::Int64,
...
     N = length(Rsets)
 
-    # allocate output
-    RsetsProj = Vector{set_type}(N)
-    oa_arg = set_type == HPolygon ? ɛ : set_type
+    # allocate output and define overapproximation function
+    if approx_sets isa Float64
+        oa = x -> overapproximate(x, HPolygon, approx_sets)
+        RsetsProj = Vector{HPolygon{numeric_type}}(N)
+    else
```
Comment (mforets) [inline src/ReachSets/project_reach.jl]: `elseif approx_sets isa LazySet` ?

Comment (schillic) [reply src/ReachSets/project_reach.jl]: From `approx_sets::Union{Float64, Type{<:LazySet}}` this case is the only one remaining, so no need to check that.

**Comment by mforets** [comment]
> The new options are called approx_init and approx_sets.

in both cases we are approximating sets. perhaps `approx_iteration`, `approx_iter`, `approx_recurrence`, `approx_inputs`?

**Comment by schillic** [comment]
I also disliked `approx_sets`, but I do not really like your candidates much more :)
Then I can change it to `approx_iter`.

**Comment by mforets** [comment]
it wasn't clear to me if in the interface to solve i can still define the iterative refinement parameter `ε=1e-3` (if not, we should add it as an alias), and that the more advanced specialization initial/iter is optional such that `solve(S, :T=>0.1, :blocks=>[1,2], :ε=>1e-3)` still works.

**Comment by schillic** [comment]
> Damn, there are random tests with random errors. I will have a look.

Those are not related to this PR, see #81.

**Comment by mforets** [comment]
for 

```julia
A = [1.04775   -1.88478   -1.95018   -0.584679
  0.224327  -0.593088  -1.58971   -1.55983 
  1.48305   -1.18984   -1.47607   -1.79205 
 -0.390503   0.979313  -0.198312  -0.774322]

X0 = BallInf(ones(4), 0.1)

@time s = solve(ContinuousSystem(A, X0), :T=>1., :N=>100, :blocks=>[1,2], :ε=>Inf, :set_type=>Hyperrectangle);
```

**Comment by mforets** [comment]
> the motor model with ε < Inf it runs forever ...

do you have the code? here this works well:

```julia
Ii = [1, 2, 2, 3, 3, 3, 3, 4, 5, 6, 6, 7, 7, 7, 7, 8]
J = [2, 3, 2, 1, 2, 3, 4, 1, 6, 7, 6, 5, 6, 7, 8, 5]
vals = [1, 8487.2, -1.0865, -2592.1, -21.119, -698.91, -141399.0, 1.0, 1.0,
       8487.2, -1.0865, -2592.1, -21.119, -698.91, -141399.0, 1.0]
A = sparse(Ii, J, vals)
A = A + sprandn(size(A)..., 0.2)

# swap variables 2 and 5
S = eye(size(A, 1))
aux = S[2, :]
S[2, :] = S[5, :]
S[5, :] = aux

X0 = Hyperrectangle([0.00225, 0.0, 0.0, 0.0, 0.00125, 0.0, 0.0, 0.0],
                    [0.00025, 0.0, 0.0, 0.0, 0.00025, 0.0, 0.0, 0.0])
B = sparse([4, 8], [1, 2], [-1.0, -1.0])
U = B * Hyperrectangle([0.23, 0.3], [0.07, 0.1])
sys = ContinuousSystem(S * A * S.', S * X0, S * U);

result = solve(sys, Options(
                            :mode => "reach",
                            :property => LinearConstraintProperty(
                                             Clause([LinearConstraint([1.; zeros(7)], 0.35),
                                                     LinearConstraint([zeros(4); 1.; zeros(3)], 0.45)])),
                                                     # x1 < 0.35 || x5 < 0.45
                            :blocks => [1, 2, 3], # blocks needed for property
                            :plot_vars => [1, 2],
                            :δ => 0.001, :N=>20, :ε=>0.001));
plot(result, alpha=0.1)
```

**Comment by schillic** [comment]
> saving the bug of hyperrectangle shape plotting, do you expect this difference on accuracy given these options?

Wow, good catch. But the options are set correctly.
Looking at X0:
```julia
julia> s.Xk[1]
LazySets.Hyperrectangle{Float64}([0.005, 1.1], [0.005, 0.861333])
julia> s2.Xk[1]
LazySets.HPolygon{Float64}(LazySets.HalfSpace{Float64}[
    LazySets.HalfSpace{Float64}([1.0, 0.0], 0.01),
    LazySets.HalfSpace{Float64}([0.0, 1.0], 1.1),
    LazySets.HalfSpace{Float64}([-1.0, 0.0], 0.0), 
    LazySets.HalfSpace{Float64}([0.0, -1.0], -0.861333)
])
```
Maybe there is a bug in `decompose` and/or `overapproximate`...

**Comment by schillic** [comment]
> do you have the code? here this works well:

Okay, my time step was just too big. We should really adapt those default settings for the models to fix `δ` and not `T`.

**Comment by schillic** [comment]
From my side this is ready.

**Pull request merged**